### PR TITLE
ssh to vm

### PR DESCRIPTION
**Pull request by metalice**
Signed-off-by: Matan Schatzman <mschatzm@redhat.com>

### CODE REVIEW

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx [hunk 1]
```diff
@@ -62,7 +62,7 @@ import { NetworkingTab } from './tabs/networking-tab/networking-tab';
 import { ReviewTab } from './tabs/review-tab/review-tab';
 import { ResultTab } from './tabs/result-tab/result-tab';
 import { StorageTab } from './tabs/storage-tab/storage-tab';
-import { CloudInitTab } from './tabs/cloud-init-tab/cloud-init-tab';
+import AdvanceTab from './tabs/advanced-tab/AdvanceTab';
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx:65]: change `AdvanceTab` to `AdvancedTab`

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts [hunk 2]
```diff
@@ -64,10 +66,12 @@ export const createVMAction = (id: string) => (dispatch, getState) => {
     .catch((error) =>
       cleanupAndGetResults(enhancedK8sMethods, error, { prettyPrintPermissionErrors: false }),
     )
-    .then(({ isValid, ...tabState }: ResultsWrapper) =>
+    .then(({ isValid, ...tabState }: ResultsWrapper) => {
+      enableSSHService &&
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts:70]: please create an if block here

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts [hunk 3]
```diff
@@ -29,6 +30,7 @@ export const createVMAction = (id: string) => (dispatch, getState) => {
 
   const state = getState();
 
+  const enableSSHService = state?.plugins?.kubevirt?.authorizedSSHKeys?.enableSSHService;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts:33]: please write a selector

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import './advanced-tab.scss';
+
+type AdvanceTabProps = {
+  wizardReduxID: string;
+  key: string;
+};
+
+const CLOUD = 'cloud';
+const SSH = 'ssh';
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx:22]: please move these constants to a consts file. maybe `/utils/strings`

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+const AdvanceTab: React.FC<AdvanceTabProps> = ({
+  wizardReduxID,
+  key,
+}: AdvanceTabProps): React.ReactElement => {
+  const { t } = useTranslation();
+  const [expended, setExpended] = React.useState<string>();
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx:29]: change to expanded

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/advanced-tab.scss [hunk 6]
```diff
@@ -0,0 +1,8 @@
+.AdvancedTab-tab-title {
+  .pf-c-accordion__toggle-text {
+    display: flex;
+    .pf-c-form__group-label-help {
+      margin-left: 6px;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/advanced-tab.scss:5]: we should not use exact numbers in CSS.
try `var(--pf-global--spacer--md)` or `var(--pf-global--spacer--lg)`

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/ssh/SSHAdvancedTab.tsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
+import SSHCreateService from '../../../../ssh-service/SSHCreateService/SSHCreateService';
+import { Stack, StackItem } from '@patternfly/react-core';
+
+const SSHAdvancedTab = () => {
+  return (
+    <Stack hasGutter>
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/ssh/SSHAdvancedTab.tsx:7]: Why do we need a `Stack` with only one `StackItem`?

File: frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx [hunk 8]
```diff
@@ -366,15 +379,33 @@ export const CreateVM: React.FC<RouteComponentProps> = ({ location }) => {
...
+                formIsValid={state.isValid && isValidSSHKey}
                 isCreating={isCreating}
                 createError={createError}
                 cleanError={() => setCreateError(undefined)}
                 onFinish={async () => {
                   resetError();
                   setCreating(true);
                   try {
-                    await createVM(state.template, sourceStatus, bootState, state, scConfigMap);
+                    const vm = await createVM(
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx:390]: can we move the whole `onFinish` block outside of the JSX

File: frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx [hunk 9]
```diff
@@ -160,6 +164,15 @@ export const CreateVM: React.FC<RouteComponentProps> = ({ location }) => {
...
   const [bootState, bootDispatch] = React.useReducer(bootFormReducer, initBootFormState);
+  const {
+    enableSSHService,
+    tempSSHKey,
+    isValidSSHKey,
+    createOrUpdateSecret,
+    updateSSHKeyInGlobalNamespaceSecret,
+    restoreDefaultSSHSettings,
+  } = useSSHKeys();
+  const { createOrDeleteSSHService } = useSSHService();
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx:175]: nice

File: frontend/packages/kubevirt-plugin/src/components/create-vm/dev-console/dev-console-create-vm-form.tsx [hunk 10]
```diff
@@ -215,7 +227,19 @@ export const DevConsoleCreateVmForm: React.FC<RouteComponentProps> = () => {
                                   undefined,
                                   state,
                                   scConfigMap,
+                                  tempSSHKey,
                                 );
+                                if (vm) {
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/dev-console/dev-console-create-vm-form.tsx:232]: same here, move the `onClick` outside of the JSX

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/SSHFormKey/ssh-form-key.scss [hunk 11]
```diff
@@ -0,0 +1,33 @@
+.SSHFormKey-show-text-area {
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/SSHFormKey/ssh-form-key.scss:1]: @matthewcarleton can you review the CSS?
we should not use these exact number values.

Can't you use a PF componenst instead @metalice ?

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/SSHFormSave/ssh-form-save-in-namespace.scss [hunk 12]
```diff
@@ -0,0 +1,7 @@
+.SSHFormSaveInNamespace-checkbox.pf-c-check {
+  margin-right: 0;
+}
+
+.SSHFormSaveInNamespace-main {
+  margin-top: 6px;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/SSHFormSave/ssh-form-save-in-namespace.scss:6]: same here, use the vars I mentioned above

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/ssh-form.scss [hunk 13]
```diff
@@ -0,0 +1,4 @@
+.SSHForm-main {
+  display: inline-flex;
+  justify-content: space-between;
+}
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/ssh-form.scss:4]: missing line ending

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHPopover/ssh-popover.scss [hunk 14]
```diff
@@ -0,0 +1,9 @@
+.SSHPopover-main {
+  .pf-c-popover__body {
+    padding-right: 10px;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHPopover/ssh-popover.scss:3]: use vars I mentioned

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHPopover/ssh-popover.scss [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+  }
+}
+
+.SSHPopover-button {
+  align-self: center;
+}
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHPopover/ssh-popover.scss:9]: line ending

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHWizard/ssh-wizard.scss [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+.SSHWizard-enable-ssh-title {
+  display: flex;
+  align-items: center;
+}
+
+.SSHWizard-authorized-key-disabled {
+  opacity: 0.5;
+  cursor: not-allowed;
+}
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHWizard/ssh-wizard.scss:18]: line ending

File: frontend/packages/kubevirt-plugin/src/components/vms/menu-actions.scss [hunk 17]
```diff
@@ -1,3 +1,13 @@
...
+
+.kubevirt-menu-actions__secondary-title {
+  margin-right: var(--pf-global--spacer--sm);
+  font-size: var(--pf-global--FontSize--xs);
+}
+
+.CopySSHCommand-disabled {
+  cursor: not-allowed;
+  opacity: 0.5;
+}
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/vms/menu-actions.scss:13]: line ending

File: frontend/packages/kubevirt-plugin/src/hooks/use-ssh-keys.ts [hunk 18]
```diff
@@ -1,74 +1,75 @@
...
-
-  const updateSSHKey = React.useCallback(
-    (sshKey?: string) => {
-      dispatch(sshActions[SSHActionsNames.updateKey](namespace, sshKey));
-    },
-    [dispatch, namespace],
+    [namespace, updateSSHKey, updateSSHTempKey],
   );
 
   React.useEffect(() => {
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/hooks/use-ssh-keys.ts:51]: should we use `useMemo` instead here?

**Review by glekner — CHANGES_REQUESTED** [review]
1st review, nice job @metalice

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts (hunk 2 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts:70]: why an if is preferred? :)

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx (hunk 5 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/AdvanceTab.tsx:29]: ?

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/ssh/SSHAdvancedTab.tsx (hunk 7 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/advanced-tab/ssh/SSHAdvancedTab.tsx:7]: In 4.9 this file will get much bigger and with more content so its ready for it

File: frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx (hunk 8 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx:390]: not really outside of jsx as we are getting data from hooks, but can be extracted to a function or a custom hook

File: frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/ssh-form.scss (hunk 13 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/ssh-service/SSHForm/ssh-form.scss:4]: is it critical? eslint isn't auto fixing/erroring about it and vs code save like this :)

File: frontend/packages/kubevirt-plugin/src/hooks/use-ssh-keys.ts (hunk 18 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/hooks/use-ssh-keys.ts:51]: not really.. useEffect will run after the rendering which is ok

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts (hunk 2 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/main-actions/create-vm.ts:71]: its more readable

File: frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx [hunk 19]
```diff
@@ -39,9 +39,13 @@ import { useStorageClassConfigMap } from '../../hooks/storage-class-config-map';
 import { createVM } from '../../k8s/requests/vm/create/simple-create';
 import { useErrorTranslation } from '../../hooks/use-error-translation';
 import { useVmTemplatesResources } from './hooks/use-vm-templates-resources';
+import { isEmpty } from 'lodash';
+import useSSHKeys from '../../hooks/use-ssh-keys';
+import useSSHService from '../../hooks/use-ssh-service';
 
 import '../create-vm-wizard/create-vm-wizard.scss';
 import './create-vm.scss';
+import { AUTHORIZED_SSH_KEYS } from '../ssh-service/SSHForm/ssh-form-utils';
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx:48]: put this import above the css onces

File: frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx (hunk 8 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/create-vm/create-vm.tsx:390]: fine leave it

**Pull request merged**